src/session.rs

```diff
@@ -0,0 +1,73 @@
+use crate::{
+    constants::{algorithm::HashingAlgorithm, types::session::SessionType},
+    handles::SessionHandle,
+    Error, Result, WrapperErrorKind,
+};
+use log::error;
+use std::convert::{From, TryFrom};
+
+/// Enum representing the different types of sessions.
+#[derive(Debug, Copy, Clone, Eq, PartialEq)]
+pub enum Session {
+    Hmac(HmacSession),
+    Policy(PolicySession),
+    Trial(TrialSession),
+}
```

Comment:
Something like this:
```rust
pub enum Session {
    Hmac{ handle: SessionHandle, auth_hash: HashingAlgorithm  },
    Policy{ handle: SessionHandle, auth_hash: HashingAlgorithm },
    Trial{ handle: SessionHandle, auth_hash: HashingAlgorithm},
}

impl Session {
    fn handle(&self) -> SessionHandle {
        match self {
            SessionHandle::Hmac(handle, _) => handle,
            SessionHandle::Policy(handle, _) => handle,
            SessionHandle::Trial(handle, _) => handle,
        }
    }

    fn auth_hash(&self) -> HashingAlgorithm {
        match self {
            SessionHandle::Hmac(_, auth_hash) => auth_hash,
            SessionHandle::Policy(_, auth_hash) => auth_hash,
            SessionHandle::Trial(_, auth_hash) => auth_hash,
        }
    }

    fn session_type(&self) -> SessionType {
        match self {
            SessionHandle::Hmac(_, _) => SessionType::Hmac,
            SessionHandle::Policy(_, _) => SessionType::Policy,
            SessionHandle::Trial(_, _) => SessionType::Trial,
        }
    }
}
```

Or did you think that we should keep the types holding data for each session type?

Reply:
That would actually work! Not necessarily what I was intending, initially I thought we could do a 
```rust
fn handle(&self) -> SessionHandle 
    match self {
        Session::Hmac(session_obj) |
        Session::Policy(session_obj) |
        Session::Trial(session_obj) => session.handle()
    }
}
```

But I was also wondering if we could get away more easily by implementing the "types" straight into the main enum like you just wrote. This removes the possibility of forcing users to provide, say, a `TrialPolicy` to one particular call, but yeah... they both have advantages :)

Reply:
The question is are there any calls that is restricted to a specific SessionType? Because if there is not then there is no point of keeping the specific session types.

Reply:
I think I lean towards keeping the specific types and expose handle, auth hash and type from the Session enum. I think that gives more flexibility if we one day come across a case were only one single session type works.

Reply:
fixed